Add support for runtime fields

javanna opened the pull request:
This PR includes the work that has been done on the runtime fields feature branch until now. The high level tasks are listed in #59332. The tasks that have not yet been completed can be worked on after merging the feature branch to master.

We are adding a new x-pack plugin called `runtime-fields` that plugs in a custom mapper which allows to define runtime fields based on a script.
The changes included in this PR that were made outside of the `x-pack/plugin/runtime-fields` directory are minimal and revolve around 1) making the `ScriptService` available while parsing index mappings so that the scripts associated to runtime fields can be compiled 2) sharing code to manipulate ranges etc. as it can be reused in runtime fields .

javanna committed:
Add runtime fields plugin under x-pack

nik9000 committed:
Hack together some script contexts

Put together some example script contexts for scripted fields. This is
almost certainly wrong, but it gets us something to think about.

nik9000 committed:
Spotless and maybe tests?

nik9000 committed:
Scripted fields: rework script contexts (#58342)

This reworks the script contexts to support multiple values for each
hit. Just like before, these are almost certainly not the final
implementation, but they give us something to iterate on.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Fixup test case after merge

Painless extensions work differently now.

nik9000 committed:
Get runtime fields tests passing

Uses a temporary hack to allow us to use painless in unit tests.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Scripted keyword field (#58939)

This is the first draft implementation of scripted field of type keyword. Still lots of TODOs to address, this PR targets a feature branch.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
adapt to upstream changes

javanna committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Add term query for keyword script fields (#59372)

This adds what I think is just about the simplest possible `term` query
implementation for `keyword` script fields and wires it into the field
mapper that we build for them.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Make ScriptFieldMapper a parameterized mapper (#59391)

Relates to #59332

nik9000 committed:
Add tests for keyword script field's fielddata (#59523)

Adds tests for the `keyword` typed `script` field's fielddata
implementation. One unit test which runs quickly and an integretion test
for fetching and an integration test for the `terms` agg.

nik9000 committed:
Two queries for keyword script field (#59527)

This adds the `exits` and `terms` query for the `keyword` style of
scripted field.

nik9000 committed:
Remaining queries for script keyword fields (#59630)

Adds the remaining queries for scripted keyword fields and handles a few
leftover TODOs, mostly around `hashCode` and `equals`.

nik9000 committed:
Allow runtime fields to depend on one another (#59636)

This updates a test so that runtime fields depend on one another and
fixes support for it.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Runtime fields: rework script service injection (#59659)

Instead of relying no createComponents to get the script service, we rely on the query shard context supplier that's available in the parser context, yet we only carry around as callback that allows us to compile the script

nik9000 committed:
Mark all scripted field queries as expensive (#59658)

This will cause them to fail if you don't have permission to execute
expensive queries.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Temporarily awaits fix failing script mapper unit tests

There seems to be no longer a way to require a field in mapper paramaters. This needs to be fixed upstream.

javanna committed:
remove leftover printStackTrace

javanna committed:
Fix parsing unsupported runtime types error

The error was calling toString on the parameter object. we need to instead print out the returned value by calling getValue

javanna committed:
Scripted keyword field type: update family type and test field caps output (#59672)

Relates to #59332

javanna committed:
Pass script params through to fielddata impl (#59762)

This addresses a TODO around using the script params, which are now parsed from the mappings. It also expand existing tests to verify that params are carried around and accessible in script for both fielddata and queries.

Relates to #59332

nik9000 committed:
Add long flavored script field (#59721)

This adds the `long` runtime type to `script` fields and implements doc
values, field data, the `term` and `exists` query.

nik9000 committed:
Render script params in script field xcontent (#59813)

This fixes the rendering of the `script` field in the `script` field
type. Originally we weren't rendering the `params` and such because we
couldn't parse them. But we can parse them now so we should render them.

This updates the integration tests to include params and adds an
integration test that fetches the mapping.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
adapt to upstream changes

javanna committed:
Fielddata implementations to not implement IndexComponent

Relates to #59724

nik9000 committed:
Remaining queries for long script field (#59816)

Adds the `terms` and `range` queries for `long` typed `script` fields.
It also fixes a bug in a few of `toString` tests for `keyword` typed
`script` fields.

javanna committed:
ScriptLongFieldData to extend LeafLongFieldData (#59884)

Relates to #59332

javanna committed:
Remove unused doXContentBody impl from ScriptFieldMapper (#59969)

Relates to #59332

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Update doc values after merge

javanna committed:
Make full text queries work against keyword scripted fields (#59990)

The keyword scripted field should apply the keyword analyzer at all times, by providing it as part of TextSearchInfo to the base mapped field type class. That is because the field does not support normalization, and if we don't force the keyword analyzer the default search analyzer will get picked up which will lowercase the query etc.

Relates to #59332

nik9000 committed:
Add double script fields (#59933)

This adds support for `double` typed `script` fields.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Adapt to upstream changes

nik9000 committed:
Replace script unit tests with integration tests (#60027)

This replaces the fancy unit tests that we were doing to test our
painless integration with an extension to our existing integration
tests. The integration tests are "more real", shorter, and are a nice
step down the road of dropping our dependency on the painless plugin in
favor of a dependency only on it's SPI.

nik9000 committed:
Drop some yml tests for script fields (#60079)

We have unit tests for all of these things and the other yml tests offer
good proof that these things work.

nik9000 committed:
Standardize script field's rejection error (#60029)

This standardizes the exception type thrown when attempting to build a
query on a script field that it doesn't support. It also replaces the
field type `[script]` in the error message with whatever the runtime type
is.

javanna committed:
Rename script field type to runtime_script (#60083)

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Replace painless unit tests with mock scripts (#60094)

Now that we have good yml tests for runtime_script fields we don't need
painless in our unit tests at all. This replaces painless with a mock
script which runs more quickly and isolates us from painless's
internals.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
adapt to upstream changes

javanna committed:
Pass a SearchLookup supplier through to fielddataBuilder (#60224)

Runtime fields depend on SearchLookup to be able to look up other fields and manipulate them. With this change we add a SearchLookup argument to the existing MappedFieldType#fielddataBuilder method. Additionally the use of runtime fields in index sorting is disallowed.

Relates to #59332

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Add runtime_script date field (#60092)

Adds support for `runtime_script` fields with `runtime_type: date`.
Doesn't add support for the `format` parameter to the mapper but *does*
support it on aggregations and the like. Its a start!

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
adapt to upstream changes

nik9000 committed:
Error on bad shape relations in runtime fields (#60463)

Most core types throw back an error if the `relation` on a `range` query
is `disjoint`. This make runtime fields do the same.

javanna committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Format support for script doc fields (#60465)

Adds `format` and `locale` support to `runtime_script` fields,
specifically those with the `runtime_type` of `date`. Others
`runtime_type`s will return an error if provided with those parameters.

javanna committed:
Runtime script field mapper to reject copy_to and fields (#60580)

Multi fields and copy_to are not supported by the runtime script field mapper. They are now rejected when provided.

nik9000 committed:
Fix build

Looks like we lost a method that I was using in a test when merging
master. Oops! This fixes the build and lets us move forward.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Implement runtime script ips (#60533)

This implements the `ip` typed runtime fields. They share a fair bit
with `string` runtime fields but we represent them as a `BytesRef`
containing 128 bits so that the comparisons all happen in the same way
as Lucene's `InetAddressPoint`.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Add the ScriptService to the field parser config (#60933)

This makes parsing runtime fields much simpler.

We'd initially resisted doing this but all of our other integration options
proved more troublesome.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Run core's integration tests with runtime fields (#60931)

Adds the `x-pack/plugin/runtime-fields/qa/rest` project to run core's
integration tests against an index with `runtime_scipt` fields. This
works by modifying the test configuration on load to replace supported
field types with `runtime_script`.

Relates to #59332

nik9000 committed:
Some progress on failing runtime fields tests (#61098)

This breaks apart the a test for the `terms` aggregation into one that
work for runtime fields and one that doesn't.

nik9000 committed:
Implement distance_feature for runtime dates (#60851)

This implements the distance_feature for `date` valued `runtime_script`s.
This produces the same numbers running against an indexed date, but it
doesn't have the same performance characteristics at all. Which is normal
for `runtime_script`s`. But `distance_feature` against an indexes fields
does a lot of work to refine the query as it goes, limiting the number of
documents that it has to visit. We can't do that because we don't have an
index. So we just spit out the same numbers and hope it is good enough.

nik9000 committed:
Borrow more tests for runtime fields (#61113)

This "borrows" 150 more tests from core for runtime fields, extending
the work done in #60931. More precisely, it adds a template to every
test that forces dynamic mapping updates to build runtime fields where
possible. In particular, `long` and `double` field are created as
runtime fields. `string`-typed fields are mimick the out of the box
behavior and create a top level `text` field with a `.keyword`
multi-field, but this `keyword` multi-field executes a script and loads
from source.

nik9000 committed:
Include boost in `distance_feature` query equals

Add `boost` to a scripted-`date`'s `distance_feature` query's equals
method.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Add boolean values script fields (#60830)

This adds `boolean` values `runtime_script` fields.

gaobinlong committed:
Fix wrong pipeline name in debug log (#58817)

nik9000 committed:
Add a consistent way to parse dates (#61105)

This adds a `parse` method to `date` valued `runtime_script`s that
converts any `Object` into a `long` using the same mechanism that
indexing a `date` uses.

We also use this new method to expand our test that run's core's
integration tests against runtime fields to `date` values fields.
Previously we couldn't do that because we didn't have a consistent way
to parse the field.

Commit:
[DOCS] Update CCR docs to focus on Kibana (#60555)

* First crack at rewriting the CCR introduction.

* Emphasizing Kibana in configuring CCR (part one).

* Many more edits, plus new files.

* Fixing test case.

* Removing overview page and consolidating that information in the main page.

* Adding redirects for moved and deleted pages.

* Removing, consolidating, and adding redirects.

* Fixing duplicate ID in redirects and removing outdated reference.

* Adding test case and steps for recreating a follower index.

* Adding steps for managing CCR tasks in Kibana.

* Adding tasks for managing auto-follow patterns.

* Fixing glossary link.

* Fixing glossary link, again.

* Updating the upgrade information and other stuff.

* Apply suggestions from code review

* Incorporating review feedback.

* Adding more edits.

* Fixing link reference.

* Adding use cases for #59812.

* Incorporating feedback from reviewers.

* Apply suggestions from code review

* Incorporating more review comments.

* Condensing some of the steps for accessing Kibana.

* Incorporating small changes from reviewers.

Co-authored-by: debadair <debadair@elastic.co>
Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

original-brownbear committed:
Fix testConcurrentSnapshotDeleteAndDeleteIndex (#61228)

There is a corner case here in which during partial snapshot the index is
deleted right between starting the snapshot in the CS and the data node getting to work
on it, causing the data node the fail that shard snapshot and making the snapshot `PARTIAL`.

Closes #61208

original-brownbear committed:
Remove Unused ByteBufferReference (#61116)

We only work with heap byte buffers at this point and those we can and do unwrap the
`byte[]` ourselves and use `BytesArray` instead of a needless level of indirection via `ByteBuffer`.

astefan committed:
Introduce ordering in the constant_keyword test for better predictibility. (#61248)

pugnascotia committed:
Add UBI docker builds (#60742)

This PR resurrects support for building Docker images based on one of Red Hat's UBI images. It also adds support for
running the existing Docker tests against the image. The image is named `elasticsearch-ubi8:<version>`.

I also changed the Docker build file uses enums instead strings in a lot of places, for added rigour.

szabosteve committed:
[DOCS] Replaces ml.node with node.roles: [ ml ] in ML settings (#61017)

nik9000 committed:
Fix `value` method on ip scripts  (#61230)

Fixes `doc['ip'].value` when `ip` is a script field. It didn't work
properly before because it wasn't whitelisted. We can't whitelist it
easilly because it was over in a module rather than core. This fixes
that by sharing more code with core that *is* whitelisted.

nik9000 committed:
Opt date valued script fields out of rate limit (#61238)

This excuses script fields from the script compilation rate limiting.
It'd be fairly sad to prevent mapping updates because of the rate limit.
And we don't expect folks to add a zillion fields. On the other hand,
once we allow these scripts on the `_search` request we might indeed
want them to be considered in the limit. But we don't support that yet
and we can deal with that when we get there.

nik9000 committed:
Drop compile limit on runtime fields scripts (#61297)

This opts all of the script contexts used by `runtime_script` our of the
compilation rate limiting because it'd be lame to prevent a mapping
update because we can't compile the script.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Revert visibility change to DocLookup constructor

This can remain package private

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

nik9000 committed:
Cleanup after merge to feature/runtime_fields

nik9000 committed:
Drop single node test case from runtime fields (#61241)

This replaces the `ESSingleNodeTestCase` with `ESTestCase` in runtime
fields, proving that we can use `ESTestCase` to test field mappers.

nik9000 committed:
Convert double script to return array (#61504)

This replaces the value collection method in `double` valued runtime
script fields with simply returning a `double[]`. Painless has some
"convert" features that allow us to define automatic conversions from
things like `double` and `Collection` into `double[]` so we use those.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
remove test that was moved and re-added while merging by mistake

nik9000 committed:
Test async and graph against runtime fields (#61651)

This runs the async search and graph tests against runtime fields.

nik9000 committed:
Revert "Convert double script to return array (#61504)"

This reverts commit 6d8170c6a45d07b2658fe37659afd5fb8793fd5a. We've
decided to stick with the `value` style functions for runtime fields.

nik9000 committed:
Standardize runtime field emit methods (#61752)

This standardizes that name for all of the "emit value" methods used by
runtime field scripts to `emitValue`. To make dates possible this adds a
a `toEpochMilli` method to convert from `TemporalAccessor` into millis
since epoch.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

javanna committed:
Adapt to upstream changes

Possilby after #61457, the priority we use for the index template that converts core yaml tests to use runtime fields is already taken, which causes an error as there ma be multiple matching templates with same priority.

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna commented on x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java:
```diff
@@ -239,7 +239,7 @@ public void execute(ClientYamlTestExecutionContext executionContext) throws IOEx
                 Map.ofEntries(
                     Map.entry("index_patterns", "*"),
-                    Map.entry("priority", Integer.MAX_VALUE),
+                    Map.entry("priority", Integer.MAX_VALUE - 1),
```
@nik9000  this does the job but I am not super proud of it. I guess this is caused by our matching pattern `*` but we have no way to say "all indices but the system ones", and I am not even sure that would work either. Thoughts?

nik9000 replied:
What was wrong? The tests were passing for me without it. I don't mind it - I just don't understand it yet.

javanna replied:
yea this came from merging master in today.  see https://github.com/elastic/elasticsearch/pull/61776/commits/acfaccb715f714ebe92b478f9dd3cdfbe7b40228

nik9000 replied:
Gotcha. It makes sense. It's kind of a shame we don't override the internal templates too to prove that all of those also work with runtime fields but its cool.

nik9000 commented on server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java:
```diff
@@ -210,6 +210,7 @@ private AbstractDistanceScoreFunction parseVariable(String fieldName, XContentPa
 
         // dates and time and geo need special handling
         parser.nextToken();
+        // TODO these ain't gonna work with runtime fields
```
I think I forgot to run these down. I'll add it to my list. But I don't think it blocks merging. We may just never need these to be honest.

javanna replied:
I don't think it is worth the effort. I reviewed the leftover TODOs and I am fine with this one staying there.

nik9000 replied:
++

nik9000 commented on x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -0,0 +1,20 @@
+ */
+
+
+/* TODO this exists so that we can run painless in unit tests which we
```
I think we can remove this. I'll add it to my list of things today.

nik9000 replied:
It also shouldn't block merging.

javanna replied:
thanks

nik9000 replied:
I pushed 34e576fd04cdcfa08fea1ddef3a65f34a142917d to drop it.

nik9000 reviewed (approved):
I recognize this code!

Like you told me yesterday, most of the stuff outside of the plugin is about plumbing the script service and handling ranges. Thanks to all of your hard work getting the rest of what we needed into master already.

I found a couple of small things to add to my personal TODO list that I'd forgot about but none of those should block merging.

LGTM

nik9000 commented:
Looks like there are some core tests failing with runtime fields:
```
08:51:41   2> REPRODUCE WITH: ./gradlew ':x-pack:plugin:runtime-fields:qa:rest:yamlRestTest' --tests "org.elasticsearch.xpack.runtimefields.rest.CoreTestsWithRuntimeFieldsIT.test {yaml=async_search/20-with-poin-in-time/Async search with point in time}" -Dtests.seed=5BF1520839E0DD82 -Dtests.security.manager=true -Dtests.locale=vi -Dtests.timezone=Asia/Anadyr -Druntime.java=11 
```

I'm cool skipping those for now to unblock this.

nik9000 committed:
Support format on runtime numbers (#61787)

This adds support for the `format` parameter on `long` and `double`
valued long fields. It only adds support for it in the request - in the
mapping it is not yet supported.

nik9000 committed:
Add API test dependency

Runtime fields needs to see the point in time api for its tests.

nik9000 committed:
Drop security policy from runtime fields

We had it for testing early on and forgot to remove it.

nik9000 commented:
> Looks like there are some core tests failing with runtime fields:

I pushed 1ea67d29f66999b0bfde4973bc37dc7d8b026c73 to fix this.

nik9000 committed:
Explicit tests for nested fields (#61761)

Adds explicit tests for nested fields to the runtime fields project.
And skip all of the core tests for nested fields. They don't really have
a chance of resolving the right fields from the source.

nik9000 committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Merge branch 'master' into feature/runtime_fields

javanna committed:
Runtime fields to expose _source and stored fields like existing scripts (#61775)

This commit exposes stored fields to runtime fields (through params._fields) and replaces the current way to access _source (source['field']) in favour of the documented way for all existing scripts: params._source .

javanna committed:
Test DLS and FLS against runtime fields (#61820)

This commit introduces integration tests for Document level security with queries against runtime fields, as well as field level security used in combination with runtime fields.

Relates to #59332

javanna committed:
Expand testing for field_caps and search across multiple indices (#61836)

This commit adds tests around field_caps and search against multiple indices, where one index has some fields defined as runtime fields, while another index has the same fields defined as concrete fields. We want to make sure that as long as the data type matches, the two can integrate seamlessly.

Relates to #59332

javanna committed:
Expand max chain depth testing (#61833)

This commit extends the coverage around max chain depth in runtime fields. We test that direct references are not counted, that 5 references are allowed while 6 are not, and we test that the same error is returned no matter where the runtime field is used.

Relates to #59332

javanna commented:
run elasticsearch-ci/docs

nik9000 committed:
Runtime field: test currentTimeMillis

Adds a test for `runtime_script` who's script contains
`currentTimeMillis`.

javanna commented:
run elasticsearch-ci/packaging-sample-windows
